File: java/scf-core/src/main/java/org/mydotey/scf/type/AbstractTypeConverter.java
package org.mydotey.scf.type;

import java.util.Objects;

/**
 * @author koqizhao
 *
 * May 22, 2018
 */
public abstract class AbstractTypeConverter<S, T> implements TypeConverter<S, T> {

    private Class<S> _sourceType;
    private Class<T> _targetType;

    public AbstractTypeConverter(Class<S> sourceType, Class<T> targetType) {
        Objects.requireNonNull(sourceType, "sourceType is null");
        Objects.requireNonNull(targetType, "targetType is null");
        _sourceType = sourceType;
        _targetType = targetType;
    }

    @Override
    public Class<S> getSourceType() {
        return _sourceType;
    }

    @Override
    public Class<T> getTargetType() {
        return _targetType;
    }

    @SuppressWarnings("rawtypes")
    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;

        if (obj == null)
            return false;

        if (getClass() != obj.getClass())
            return false;

        AbstractTypeConverter other = (AbstractTypeConverter) obj;
        return Objects.equals(_sourceType, other._sourceType) && Objects.equals(_targetType, other._targetType);
    }

    @Override
    public String toString() {
        return String.format("%s { sourceType: %s, targetType: %s }", getClass().getSimpleName(), _sourceType,
                _targetType);
    }

}

File: dotnet/src/MyDotey.SCF/Type/AbstractTypeConverter.cs
using System;
using sType = System.Type;

namespace MyDotey.SCF.Type
{
    /**
     * @author koqizhao
     *
     * May 22, 2018
     */
    public abstract class AbstractTypeConverter<S, T> : ITypeConverter<S, T>
    {
        public virtual sType SourceType { get; protected set; }
        public virtual sType TargetType { get; protected set; }

        public AbstractTypeConverter()
        {
            SourceType = typeof(S);
            TargetType = typeof(T);
        }

        public virtual object Convert(object source)
        {
            return Convert((S)source);
        }

        public abstract T Convert(S source);

        public override int GetHashCode()
        {
            int prime = 31;
            int result = 1;
            result = prime * result + typeof(S).GetHashCode();
            result = prime * result + typeof(T).GetHashCode();
            return result;
        }

        public override bool Equals(object obj)
        {
            if (object.ReferenceEquals(this, obj))
                return true;

            if (obj == null)
                return false;

            return GetType() == obj.GetType();
        }

        public override String ToString()
        {
            return string.Format("{0} {{ sourceType: {1}, targetType: {2} }}", GetType().Name, SourceType,
                TargetType);
        }
    }
}